api/eds.proto

```diff
@@ -0,0 +1,183 @@
+  HealthStatus health_status = 2;
+  // The optional canary status of the upstream host. Envoy uses the canary
+  // status for various statistics and load balancing tasks documented
+  // elsewhere.
+  google.protobuf.BoolValue canary = 3;
+  // The optional load balancing weight of the upstream host, in the range 1 -
+  // 100. Envoy uses the load balancing weight in some of the built in load
+  // balancers.
+  // TODO(htuch): How does this combine with the LocalityEndpoints weight?
+  google.protobuf.UInt32Value load_balancing_weight = 4;
+}
+
+// All endpoints belonging to a Locality.
+message LocalityEndpoints {
+  Locality locality = 1;
+  repeated LbEndpoint lb_endpoints = 2;
+
+  // Optional: Per region/zone/sub_zone weight - range 1-100.
+  // TODO(htuch): Is there some restriction that all weights in a cluster must
+  // sum to 100? Do we want to be more specific and have per-zone and
+  // per-subzone as distinct weights for example? Can the weight also be 0?
+  google.protobuf.UInt32Value load_balancing_weight = 3;
+}
+
+message ClusterEndpoints {
+  string cluster_name = 1;
+  repeated LocalityEndpoints locality_endpoints = 2;
+}
+
+message EndpointDiscoveryResponse {
+  repeated ClusterEndpoints cluster_endpoints = 1;
+}
+
+// Example load report from a single request:
+//
+// [metric name, metric value]
+// * cpu_seconds, 0.7
+// * flash_utilization, 75
+//
+// When aggregating Envoy needs to count how many request's load reports
+// included each metric type, so Envoy can account for requests that don't
+// include that metric type. e.g.:
+//
+// [name, count, sum of values]
+// * cpu_seconds, 10, 17.5
+// * flash_utilization, 5, 375
+message EndpointLoadMetricStats {
+  // Name of the metric; may be empty.
+  string metric_name = 1;
+  // Number of calls that finished and included this metric.
+  uint64 num_requests_finished_with_metric = 2;
+  // Sum of metric values across all calls that finished with this metric for
+  // load_reporting_interval.
+  double total_metric_value = 3;
+}
+
+// These are stats Envoy reports to GLB every so often. Report frequency is
+// defined by LoadBalanceResponse.interval
+// Stats per upstream region/zone and optionally per subzone
+message UpstreamLocalityStats {
+  // Name of zone, region and optionally endpoint group this metrics was
+  // collected from. Zone and region names could be empty if unknown.
+  string Locality = 1;
+
+  // The total number of requests sent by this Envoy since the last report.
+  uint64 total_requests = 2;
+  // The total number of unfinished requests
+  uint64 total_requests_in_progress = 3;
+  // The number of errors since the last report.
+  // TODO(htuch): Ideally we would have the tcp_errors key be TcpErrorType, but
+  // enums are not supported as map key types.
+  enum TcpErrorType {
+    TIMEOUT = 0;
+    // TODO(htuch): Fill in additional TcpErrorType values.
+  }
+  map<uint32, uint64> tcp_errors = 4;
```

Comment:
As mentioned in the comment, ideally it would be `TcpErrorType`, but maps don't support enums. @louiscryan is there a better way to model this? I was thinking maybe just a repeated message with `TcpErrorType` and value.